hudi-common/src/main/java/org/apache/hudi/common/table/read/HoodieFileGroupReader.java

```diff
@@ -0,0 +1,233 @@
+   */
+  public void initRecordIterators() {
+    this.baseFileIterator = baseFilePath.isPresent()
+        ? readerContext.getFileRecordIterator(
+        baseFilePath.get().getHadoopPath(), 0, baseFilePath.get().getFileLen(),
+        readerState.baseFileAvroSchema, readerState.baseFileAvroSchema, hadoopConf)
+        : new EmptyIterator<>();
+    scanLogFiles();
+  }
+
+  /**
+   * @return {@code true} if the next record exists; {@code false} otherwise.
+   * @throws IOException on reader error.
+   */
+  public boolean hasNext() throws IOException {
+    while (baseFileIterator.hasNext()) {
+      T baseRecord = baseFileIterator.next();
+      String recordKey = readerContext.getRecordKey(baseRecord, readerState.baseFileAvroSchema);
+      Pair<Option<T>, Map<String, Object>> logRecordInfo = logFileRecordMapping.remove(recordKey);
+      Option<T> resultRecord = logRecordInfo != null
+          ? merge(Option.of(baseRecord), Collections.emptyMap(), logRecordInfo.getLeft(), logRecordInfo.getRight())
+          : merge(Option.empty(), Collections.emptyMap(), Option.of(baseRecord), Collections.emptyMap());
+      if (resultRecord.isPresent()) {
+        nextRecord = readerContext.seal(resultRecord.get());
+        return true;
+      }
+    }
+
+    if (logRecordIterator == null) {
+      logRecordIterator = logFileRecordMapping.values().iterator();
+    }
+
+    while (logRecordIterator.hasNext()) {
+      Pair<Option<T>, Map<String, Object>> nextRecordInfo = logRecordIterator.next();
+      Option<T> resultRecord = merge(Option.empty(), Collections.emptyMap(),
+          nextRecordInfo.getLeft(), nextRecordInfo.getRight());
+      if (resultRecord.isPresent()) {
+        nextRecord = readerContext.seal(resultRecord.get());
+        return true;
+      }
+    }
+
+    return false;
+  }
+
+  /**
+   * @return The next record after calling {@link #hasNext}.
+   */
+  public T next() {
+    T result = nextRecord;
+    nextRecord = null;
+    return result;
+  }
+
+  private void scanLogFiles() {
+    if (logFilePathList.isPresent()) {
+      FileSystem fs = readerContext.getFs(logFilePathList.get().get(0), hadoopConf);
+      HoodieMergedLogRecordReader<T> logRecordReader = HoodieMergedLogRecordReader.newBuilder()
+          .withHoodieReaderContext(readerContext)
+          .withFileSystem(fs)
+          .withBasePath(readerState.tablePath)
+          .withLogFilePaths(logFilePathList.get())
+          .withLatestInstantTime(readerState.latestCommitTime)
+          .withReaderSchema(readerState.logRecordAvroSchema)
+          .withReadBlocksLazily(getBooleanWithAltKeys(props, HoodieReaderConfig.COMPACTION_LAZY_BLOCK_READ_ENABLE))
+          .withReverseReader(false)
+          .withBufferSize(getIntWithAltKeys(props, HoodieMemoryConfig.MAX_DFS_STREAM_BUFFER_SIZE))
+          .withPartition(getRelativePartitionPath(
+              new Path(readerState.tablePath), new Path(logFilePathList.get().get(0)).getParent()
```

Comment:
can we parameterize the partition path itself? then we can get rid of `Path` usage.